src/App/Fossa/VSI/Fingerprint.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module App.Fossa.VSI.Fingerprint (
+  fingerprintRaw,
+  fingerprintContentsRaw,
+  fingerprintCommentStripped,
+  fingerprint,
+  Fingerprint,
+  Raw,
+  CommentStripped,
+  Combined (..),
+) where
+
+import Conduit (ConduitT, Void, await, decodeUtf8C, encodeUtf8C, filterC, linesUnboundedC, mapC, runConduitRes, sourceFile, yield, (.|))
+import Control.Algebra (Has)
+import Control.Effect.Diagnostics (Diagnostics, fatalText)
+import Control.Effect.Exception (IOException, Lift, catch)
+import Control.Effect.Lift (sendIO)
+import Crypto.Hash (Digest, HashAlgorithm, SHA256 (..), hashFinalize, hashInit, hashUpdate)
+import Data.Aeson (ToJSON, object, toJSON, (.=))
+import Data.ByteString qualified as B
+import Data.String.Conversion (ToText (..))
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.Extra (breakOnAndRemove)
+import Discovery.Walk (WalkStep (..), walk')
+import Effect.ReadFS (ReadFS, contentIsBinary)
+import Path (Abs, Dir, File, Path, toFilePath)
+
+-- | Fingerprint deterministically idenfies a file and is derived from its content.
+--
+-- The type variable is:
+--
+--   @k@ - Kind, the kind of fingerprint computed.
+--
+-- For ease of implementation, the backing representation of a @Fingerprint@ instance is a @Base16@ encoded @Text@.
+newtype Fingerprint k = Fingerprint Text
```

Comment:
Let's add a type role here:
```suggestion
newtype Fingerprint k = Fingerprint Text
type role Fingerprint nominal
```

I wrote a explanation for this reasoning in the [`Data.Tagged` module](https://github.com/fossas/fossa-cli/blob/8e7777e3fb07c1632639d406b1e3c67530eae9c1/src/Data/Tagged.hs#L12) (though it's worth noting that that module is actually wrong right now).

In short, this prevents `coerce :: Fingerprint Raw -> Fingerprint CommentStripped`.